resolve.go

```diff
@@ -9,59 +9,81 @@ import (
 )
 
 var ResolvableProtocols = []ma.Protocol{DnsaddrProtocol, Dns4Protocol, Dns6Protocol, DnsProtocol}
-var DefaultResolver = &Resolver{Backend: net.DefaultResolver}
+var DefaultResolver = &Resolver{def: net.DefaultResolver}
 
 const dnsaddrTXTPrefix = "dnsaddr="
 
-type Backend interface {
+// BasicResolver is a low level interface for DNS resolution
+type BasicResolver interface {
 	LookupIPAddr(context.Context, string) ([]net.IPAddr, error)
 	LookupTXT(context.Context, string) ([]string, error)
 }
 
+// Resolver is an object capable of resolving dns multiaddrs by using one or more BasicResolvers;
+// it supports custom per domain/TLD resolvers.
+// It also implements the BasicResolver interface so that it can act as a custom per domain/TLD
+// resolver.
 type Resolver struct {
-	Backend Backend
+	def    BasicResolver
+	custom map[string]BasicResolver
 }
 
-var _ Backend = (*MockBackend)(nil)
+var _ BasicResolver = (*Resolver)(nil)
 
-type MockBackend struct {
-	IP  map[string][]net.IPAddr
-	TXT map[string][]string
+// NewResolver creates a new Resolver instance with the specified options
+func NewResolver(opts ...Option) (*Resolver, error) {
+	r := &Resolver{def: net.DefaultResolver}
+	for _, opt := range opts {
+		err := opt(r)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return r, nil
 }
 
-func (r *MockBackend) LookupIPAddr(ctx context.Context, name string) ([]net.IPAddr, error) {
-	results, ok := r.IP[name]
-	if ok {
-		return results, nil
-	} else {
-		return []net.IPAddr{}, nil
+type Option func(*Resolver) error
+
+// WithDefaultResolver is an option that specifies the default basic resolver,
+// which resolves any TLD that doesn't have a custom resolver.
+// Defaults to net.DefaultResolver
+func WithDefaultResolver(def BasicResolver) Option {
+	return func(r *Resolver) error {
+		r.def = def
+		return nil
 	}
 }
 
-func (r *MockBackend) LookupTXT(ctx context.Context, name string) ([]string, error) {
-	results, ok := r.TXT[name]
-	if ok {
-		return results, nil
-	} else {
-		return []string{}, nil
+// WithTLDResolver specifies a custom resolver for a domain/TLD.
+func WithDomainResolver(domain string, rslv BasicResolver) Option {
```

Comment:
Maybe let's add a comment here indicating that we check resolvers from most specific to least specific (and have a similar comment in the `getResolver` function for easier code reading).

Reply:
sure, will do.

Reply:
done.